### PR TITLE
Add priority-ordered allocation queue

### PR DESCRIPTION
**Pull request by Priyankakarumuru1**
This adds a priority-ordered allocation queue (memory/allocation_queue.rs) that will be used to manage buffer allocation requests when memory is full. Instead of failing when there's no memory available, requests can now wait in the queue and get served in priority order - high priority and low priority.
  
The queue exposes `push_high`, `push_low`, `pop_front_if`, `upgrade`, `has_pending` and `fulfill`. It doesn't allocate buffers or check available memory itself, the pool will drive the wake loop by calling `pop_front_if` with a `can_allocate` predicate, then allocating and delivering the buffer. The wiring into the pool is separate work.

### Does this change impact existing behavior?

No. This is a new standalone module with no callers yet. Existing behavior is unchanged.

### Does this change need a changelog entry? Does it require a version change?

No, these are internal changes to the memory limiter with no user-visible behavior change.

---

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and I agree to the terms of the [Developer Certificate of Origin (DCO)](https://developercertificate.org/).

### CODE REVIEW

File: mountpoint-s3-fs/src/memory/allocation_queue.rs [hunk 1]
```diff
@@ -0,0 +1,332 @@
+//! Priority-ordered allocation queue for buffer requests under memory pressure.
+
+use std::collections::VecDeque;
+use std::sync::Mutex;
+use std::sync::atomic::{AtomicBool, Ordering};
```
Comment (yerzhan7) [inline mountpoint-s3-fs/src/memory/allocation_queue.rs]: I think we should import from `crate::sync` so that Shuttle tests work.

I.e.

```rs
  use crate::sync::Mutex;
  use crate::sync::atomic::{AtomicBool, Ordering};
```

File: mountpoint-s3-fs/src/memory/allocation_queue.rs [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+/// This queue does not allocate buffers or check available memory — it only manages
+/// ordering, signaling, and lifecycle. The pool drives the wake loop by calling
+/// [`pop_front_if`](Self::pop_front_if) with a predicate (e.g., `can_allocate`),
+/// then performing the allocation and delivering the buffer via the entry's sender.
+pub struct AllocationQueue {
+    /// Queue state protected by a mutex.
+    inner: Mutex<AllocationQueueInner>,
+    /// `true` if either queue has pending entries. Allows a lock-free check
+    /// via [`is_memory_pressure`](Self::is_memory_pressure).
+    memory_pressure: AtomicBool,
```
Comment (yerzhan7) [inline mountpoint-s3-fs/src/memory/allocation_queue.rs]: I think we should call it `is_empty` instead of `memory_pressure`? And then callers can decide if it means "memory pressure" or not. What do you think?

File: mountpoint-s3-fs/src/memory/allocation_queue.rs [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+use super::buffers::PoolBuffer;
+use super::stats::BufferKind;
+
+/// Priority levels for allocation requests.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum AllocationPriority {
+    /// Speculative prefetch — nobody is waiting on this data yet.
+    Low,
+    /// Active read or upload — a syscall is blocked waiting for this buffer.
+    High,
```
Comment (yerzhan7) [inline mountpoint-s3-fs/src/memory/allocation_queue.rs]: The comment leaks the caller concepts. Let's make comment for generic for callers/consumers of the queue?

```
  /// Priority for an allocation request.
  ///
  /// The queue serves all [High] entries (FIFO) before any [Low] entries (FIFO).
  /// A [Low] request can be promoted to the back of [High] via
  /// [`AllocationQueue::upgrade`].
  #[derive(Debug, Clone, Copy, PartialEq, Eq)]
  struct AllocationPriority {
```

File: mountpoint-s3-fs/src/memory/allocation_queue.rs [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+        let mut inner = self.inner.lock().unwrap();
+        let front = inner.high.front().or_else(|| inner.low.front())?;
+        if !predicate(front.size, front.kind, front.cursor_id) {
+            return None;
+        }
+        let entry = inner.high.pop_front().or_else(|| inner.low.pop_front());
+        if inner.high.is_empty() && inner.low.is_empty() {
+            self.memory_pressure.store(false, Ordering::Release);
+        }
+        entry
```
Comment (yerzhan7) [inline mountpoint-s3-fs/src/memory/allocation_queue.rs:170]: Should we return `entry` here if Receiver was cancelled? (`entry.sender.is_canceled()`, e.g. when speculative prefetch is dropped)

Or are we planning to have separate `cancel(CursorID)` API on this queue?

File: mountpoint-s3-fs/src/memory/allocation_queue.rs [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+
+    /// Moves all entries for `cursor_id` from the low-priority queue to the back
+    /// of the high-priority queue. No-op if no entries match.
+    ///
+    /// Called when a FUSE read arrives for a cursor that has pending speculative
+    /// allocations — those allocations are now urgent.
+    pub fn upgrade(&self, cursor_id: CursorId) {
+        let mut inner = self.inner.lock().unwrap();
+        let mut to_promote = Vec::new();
+        let mut remaining = VecDeque::with_capacity(inner.low.len());
```
Comment (yerzhan7) [inline mountpoint-s3-fs/src/memory/allocation_queue.rs]: This will overallocate when not strictly needed. What if just do in-place rotation?

```
let mut inner = self.inner.lock().unwrap();
let n = inner.low.len();
for _ in 0..n {
    let entry = inner.low.pop_front().unwrap();
    if entry.cursor_id == Some(cursor_id) {
        inner.high.push_back(entry);
    } else {
        inner.low.push_back(entry);
    }
}
```

File: mountpoint-s3-fs/src/memory.rs [hunk 6]
```diff
@@ -1,12 +1,12 @@
+#[allow(unused)]
+pub(crate) mod allocation_queue;
```
Comment (yerzhan7) [inline mountpoint-s3-fs/src/memory.rs]: Should it be just `mod allocation_queue;` to be consistent with modules below?

File: mountpoint-s3-fs/src/memory.rs [hunk 7]
```diff
@@ -1,12 +1,12 @@
+#[allow(unused)]
```
Comment (yerzhan7) [inline mountpoint-s3-fs/src/memory.rs:2]: Could we add TODO comment here so that we don't forget? e.g. `// TODO(memory-limiter): remove once wired into PagedPool`

File: mountpoint-s3-fs/src/memory/allocation_queue.rs [hunk 8]
```diff
@@ -0,0 +1,332 @@
...
+    /// Which cursor is requesting this buffer. `None` for upload requests
+    /// (uploads have no associated cursor).
+    pub cursor_id: Option<CursorId>,
+    /// Size of the buffer the waiter needs, in bytes.
+    pub size: usize,
+    /// The kind of buffer to allocate (e.g. [BufferKind::GetObject], [BufferKind::PutObject]).
+    pub kind: BufferKind,
+    /// Channel sender used to deliver the allocated [PoolBuffer] to the waiter.
+    /// When dropped, the receiver resolves with `Err(Canceled)`.
+    pub sender: oneshot::Sender<PoolBuffer>,
```
Comment (yerzhan7) [inline mountpoint-s3-fs/src/memory/allocation_queue.rs]: np: `sender` could be private behind a `fulfill(self, buffer) -> Result<(), PoolBuffer>` method on `PendingAllocation` struct.

File: mountpoint-s3-fs/src/memory/allocation_queue.rs [hunk 9]
```diff
@@ -0,0 +1,332 @@
...
+            kind,
+            sender,
+        };
+
+        let mut inner = self.inner.lock().unwrap();
+        match priority {
+            AllocationPriority::High => inner.high.push_back(entry),
+            AllocationPriority::Low => inner.low.push_back(entry),
+        }
+        self.memory_pressure.store(true, Ordering::Release);
```
Comment (yerzhan7) [inline mountpoint-s3-fs/src/memory/allocation_queue.rs]: np: `self.memory_pressure.store(true, Ordering::Release);` can be moved after `let mut inner = self.inner.lock().unwrap();`

File: mountpoint-s3-fs/src/memory/allocation_queue.rs [hunk 10]
```diff
@@ -0,0 +1,332 @@
...
+        assert_eq!(second.cursor_id, Some(CursorId::new_from_raw(2)));
+    }
+
+    #[test]
+    fn test_upgrade_moves_low_to_high() {
+        let queue = AllocationQueue::new();
+        let cursor_a = CursorId::new_from_raw(1);
+        let cursor_b = CursorId::new_from_raw(2);
+
+        let _rx_a = queue.push(AllocationPriority::Low, Some(cursor_a), 1024, BufferKind::GetObject);
```
Comment (yerzhan7) [inline mountpoint-s3-fs/src/memory/allocation_queue.rs]: Can we also test/assert that if we push ` cursor_id: None` (uploads) and then call `upgrade(some_cursor)` the upload request is still in the correct place in the queue?

File: mountpoint-s3-fs/src/memory/allocation_queue.rs [hunk 11]
```diff
@@ -0,0 +1,332 @@
...
+            cursor_id,
+            size,
+            kind,
+            sender,
+        };
+
+        let mut inner = self.inner.lock().unwrap();
+        match priority {
+            AllocationPriority::High => inner.high.push_back(entry),
+            AllocationPriority::Low => inner.low.push_back(entry),
```
Comment (yerzhan7) [inline mountpoint-s3-fs/src/memory/allocation_queue.rs:128]: Should we fail if we trying to push Low priority request without cursor ID? I.e. such request must always be high priority?

File: mountpoint-s3-fs/src/memory/allocation_queue.rs (hunk 10 above)
Comment (Priyankakarumuru1) [reply mountpoint-s3-fs/src/memory/allocation_queue.rs]: Good idea. Added a test for that. Thanks!

File: mountpoint-s3-fs/src/memory/allocation_queue.rs (hunk 5 above)
Comment (Priyankakarumuru1) [reply mountpoint-s3-fs/src/memory/allocation_queue.rs]: Good one.

File: mountpoint-s3-fs/src/memory/allocation_queue.rs (hunk 4 above)
Comment (Priyankakarumuru1) [reply mountpoint-s3-fs/src/memory/allocation_queue.rs:170]: For now we aren't checking for cancelled receivers here. What happens is that if a cursor is dropped while its entry is in the queue, the receiver goes away. When pop_front_if pops that entry and the pool calls fulfill(buffer), the send fails and the buffer is dropped immediately. We can add a cancel API later as an optimization if needed.

File: mountpoint-s3-fs/src/memory/allocation_queue.rs (hunk 11 above)
Comment (Priyankakarumuru1) [reply mountpoint-s3-fs/src/memory/allocation_queue.rs:128]: Good point. A low priority request without a cursor_id shouldn't happen. Low = speculative prefetch which always has a cursor. Added an assert! that low priority requests must have cursor_id.

File: mountpoint-s3-fs/src/memory/allocation_queue.rs (hunk 2 above)
Comment (Priyankakarumuru1) [reply mountpoint-s3-fs/src/memory/allocation_queue.rs]: Makes sense, I've renamed it to `has_pending` to keep it generic. I didn't go with is_empty because most callers want to know 'is there someone waiting?', with `is_empty` they'd have to write !is_empty() every time which feels a bit backwards. `has_pending` gives you a direct yes/no.

File: mountpoint-s3-fs/src/memory/allocation_queue.rs (hunk 3 above)
Comment (Priyankakarumuru1) [reply mountpoint-s3-fs/src/memory/allocation_queue.rs]: Good point, updated the comment to be generic.

File: mountpoint-s3-fs/src/memory/allocation_queue.rs (hunk 4 above)
Comment (yerzhan7) [reply mountpoint-s3-fs/src/memory/allocation_queue.rs:170]: I think we should remove/prune all canceled receivers (from the front of the queue) here, while we hold queue mutex. It's not only about wasted buffer allocations. E.g. if the front entry in the queue was cancelled (and e.g. it's of size 64MB), then the predicate (when only 8MB is free) will never actually return anything from the queue. And live low-priority 8MB request (next item after 64MB request) stays blocked in the queue.

We should remove/prune cancelled receivers in the `upgrade()` method as well while we iterate through low priority queue.

File: mountpoint-s3-fs/src/memory/allocation_queue.rs (hunk 11 above)
Comment (yerzhan7) [reply mountpoint-s3-fs/src/memory/allocation_queue.rs:128]: Instead of `assert` what if we split `push` into `push_low()` and `push_high()` methods to enforce this at compile time instead of runtime? 

```rs
  pub fn push_high(&self, cursor_id: Option<CursorId>, size: usize, kind: BufferKind)
      -> oneshot::Receiver<PoolBuffer>
  {
      self.push_inner(AllocationPriority::High, cursor_id, size, kind)
  }

  pub fn push_low(&self, cursor_id: CursorId, size: usize, kind: BufferKind)
      -> oneshot::Receiver<PoolBuffer>
  {
      self.push_inner(AllocationPriority::Low, Some(cursor_id), size, kind)
  }
```

Comment (Priyankakarumuru1) [reply mountpoint-s3-fs/src/memory/allocation_queue.rs:128]: Done

File: mountpoint-s3-fs/src/memory/allocation_queue.rs (hunk 4 above)
Comment (Priyankakarumuru1) [reply mountpoint-s3-fs/src/memory/allocation_queue.rs:170]: Good one.

**Review by yerzhan7 — APPROVED** [review]
